### PR TITLE
Use script path as location for nuget.config file

### PR DESCRIPTION
**Pull request by KevinRansom**
A fairly common request we have got for #r nuget is to allow script authors to make use of nuget.config files, using the normal nuget resolution rules.

Currently we use the standard nuget.config resolving from the temp directory where we create the temporary project file for package resolution, following this change we will gather sources from nuget.config resolved from the script directory.

The script directory is the:
- initial working directory when running in fsi.
- The directory containing the --exec script.fsx file when executing a script from disk.

Todo:
- Decide what the behaviour should be about script being edited in the IDE, and then fragments sent to FSI.  Today there is no real state alignment between the two executions, there probably should be some.

With this change it will be possible to add a Nuget.config alongside the script file and specify additional sources in that nuget.config file.  The mechanism will also probe directories above looking for specified sources as well as  the user Nuget.config file.

It should be noted that, this PR: does not enable other facilities from nuget.config files, such as authentication for private feeds.

Whilst doing this we also stopped using msbuild on the desktop to build the generated project.  And extended the probe for dotnet.exe to make use of the PATH environment variable on the desktop.

### CODE REVIEW

File: src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.Utilities.fs [hunk 1]
```diff
@@ -129,32 +92,56 @@ module internal Utilities =
...
+                if File.Exists(pathToFile) then 
+                    true
+                else
+                    false
+            with | _ -> false
+        // Use the path to search for dotnet.exe
+        let probePathForDotnetHost () =
+            let paths =
+                let p = Environment.GetEnvironmentVariable("PATH")
+                if not(isNull p) then p.Split(';')
```
Comment (brettfo) [inline src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.Utilities.fs:106]: Does this code path get hit on non-Windows?  You'll need to split on `Path.DirectorySeparatorChar` and `Path.AltDirectorySeparatorChar`.

File: src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.Utilities.fs [hunk 2]
```diff
@@ -129,32 +92,56 @@ module internal Utilities =
...
-                    let dotnetApp =
-                        let platform = Environment.OSVersion.Platform
-                        if platform = PlatformID.Unix then "dotnet" else "dotnet.exe"
+                let candidate =
                     let assemblyLocation = Path.GetDirectoryName(typeof<Int32>.GetTypeInfo().Assembly.Location)
-                    Path.GetFullPath(Path.Combine(assemblyLocation, "../../..", dotnetApp))
-
-                if File.Exists(dotnetLocation) then
-                    Some dotnetLocation
+                    Path.GetFullPath(Path.Combine(assemblyLocation, "../../..", dotnet))
```
Comment (brettfo) [inline src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.Utilities.fs:118]: You can pass multiple args to `Combine()`, then you don't have to worry about the separator.

```suggestion
                    Path.GetFullPath(Path.Combine(assemblyLocation, "..", "..", "..", dotnet))
```

File: src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.Utilities.fs [hunk 3]
```diff
@@ -129,32 +92,56 @@ module internal Utilities =
...
                         Some main.FileName
                     else
                         Some dotnet
+        elif isWindows then
+            // Use the path to search for dotnet.exe
+            match probePathForDotnetHost () with
+            | Some f -> Some (Path.Combine(f, dotnet))
+            | None ->
+                // Not in paths so look for global install of dotnet sdk
+                let pf = Environment.GetEnvironmentVariable("ProgramW6432")
```
Comment (brettfo) [inline src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.Utilities.fs:133]: What about 32-bit Windows?  This variable isn't present.

File: src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.Utilities.fs (hunk 1 above)
Comment (KevinRansom) [reply src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.Utilities.fs:106]: Aren't they the '\' and '/' ?

File: src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.Utilities.fs (hunk 3 above)
Comment (brettfo) [reply src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.Utilities.fs:133]: JK, you handle it on the next line.

File: src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.Utilities.fs [hunk 4]
```diff
@@ -237,3 +219,24 @@ module internal Utilities =
...
+#endif
+        seq {
+            if success then
+                for source in stdOut do
+                    // String returned by dotnet nuget list source --format short
+                    // is formatted similar to:
+                    // E https://dotnetfeed.blob.core.windows.net/dotnet-core/index.json
+                    // So strip off the flags
+                    let pos = source.IndexOf(" ")
+                    if pos >= 0 then yield ("i", source.Substring(pos).Trim(' '))
```
Comment (brettfo) [inline src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.Utilities.fs:241]: Might as well make sure there aren't any newlines, either.

```suggestion
                    if pos >= 0 then yield ("i", source.Substring(pos).Trim())
```

File: src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.Utilities.fs (hunk 1 above)
Comment (KevinRansom) [reply src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.Utilities.fs:106]: This is about splitting the elements of the path environment variable.

Comment (brettfo) [reply src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.Utilities.fs:106]: Japanese systems have a different character, I think the Yen symbol.

Comment (KevinRansom) [reply src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.Utilities.fs:106]: Hmm it's colon on linux not semi colon.

Comment (brettfo) [reply src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.Utilities.fs:106]: I make this mistake EVERY SINGLE TIME.

What I should have said is `Path.PathSeparator` which is ';' on Windows and ':' on Unix.

**Review by TIHan — APPROVED** [review]

**Review by brettfo — APPROVED** [review]

**Pull request merged**